File: packages/cli/src/commands/import.ts
import { Command } from "commander";
import chalk from "chalk";
import ora from "ora";
import { createApiClient } from "../lib/api-client.js";
import { handleError, validateDateFormat } from "../lib/shared.js";
import {
  ALL_PARSERS,
  detectAvailableIDEs,
  getParser,
  conversationToEvents,
  type IDEParser,
  type IDESource,
  type IDEProject,
  type ParsedConversation,
} from "../lib/ide-parsers/index.js";

/* ─── Constants ─── */
const BATCH_SIZE = 20;
const MAX_EVENTS_PER_REQUEST = 100;

/* ─── remb import ─── */

export const importCommand = new Command("import")
  .description("Import AI chat history from local IDE storage into Remb")
  .option("--ide <name>", "Import from a specific IDE only (e.g., cursor, claude-code, vscode)")
  .option("--project <id>", "Import a specific project/workspace by its storage ID")
  .option("--remb-project <slug>", "Associate imported conversations with this Remb project")
  .option("--all", "Import all conversations from all detected IDEs without prompting")
  .option("--dry-run", "Show what would be imported without actually sending data")
  .option("--since <date>", "Only import conversations after this date (YYYY-MM-DD)")
  .option("--list", "List detected IDEs and available projects, then exit")
  .option("-l, --limit <n>", "Max conversations to import per project", "100")
  .addHelpText(
    "after",
    `
Supported IDEs:
  cursor          Cursor (VS Code fork)
  claude-code     Claude Code CLI
  vscode          VS Code (GitHub Copilot)
  windsurf        Windsurf (Codeium)
  intellij        IntelliJ IDEA
  pycharm         PyCharm
  android-studio  Android Studio
  visual-studio   Visual Studio (Windows)
  zed             Zed
  sublime-text    Sublime Text (LSP-Copilot)

Examples:
  $ remb import                          # Auto-detect and import interactively
  $ remb import --list                   # List all detected IDEs and projects
  $ remb import --ide cursor             # Import from Cursor only
  $ remb import --ide cursor --project abc123  # Import specific workspace
  $ remb import --all --dry-run          # Preview what would be imported
  $ remb import --since 2025-01-01       # Only import recent conversations
  $ remb import --ide claude-code --remb-project my-app  # Associate with Remb project`,
  )
  .action(async (opts) => {
    try {
      await runImport(opts);
    } catch (err) {
      handleError(err);
    }
  });

/* ─── Main import logic ─── */

interface ImportOpts {
  ide?: string;
  project?: string;
  rembProject?: string;
  all?: boolean;
  dryRun?: boolean;
  since?: string;
  list?: boolean;
  limit?: string;
}

async function runImport(opts: ImportOpts) {
  // Validate options
  if (opts.since) validateDateFormat(opts.since, "--since");
  const limit = Math.min(parseInt(opts.limit ?? "100", 10) || 100, 500);
  const sinceDate = opts.since ? new Date(`${opts.since}T00:00:00Z`) : undefined;

  // Step 1: Detect available IDEs
  const spinner = ora("Detecting installed IDEs...").start();

  let parsers: IDEParser[];
  if (opts.ide) {
    const parser = getParser(opts.ide as IDESource);
    if (!parser) {
      spinner.fail(`Unknown IDE: ${opts.ide}`);
      console.log(chalk.dim(`  Supported: ${ALL_PARSERS.map((p) => p.id).join(", ")}`));
      process.exit(1);
    }
    const available = await parser.detect();
    if (!available) {
      spinner.fail(`${parser.displayName} storage not found on this machine`);
      process.exit(1);
    }
    parsers = [parser];
  } else {
    parsers = await detectAvailableIDEs();
  }

  if (parsers.length === 0) {
    spinner.fail("No supported IDEs detected on this machine");
    process.exit(1);
  }

  spinner.succeed(`Found ${parsers.length} IDE${parsers.length > 1 ? "s" : ""}: ${parsers.map((p) => p.displayName).join(", ")}`);

  // Step 2: List projects for each IDE
  const allProjects: Array<{ parser: IDEParser; project: IDEProject }> = [];

  for (const parser of parsers) {
    const projectSpinner = ora(`  Scanning ${parser.displayName} workspaces...`).start();
    try {
      const projects = await parser.listProjects();

      if (opts.project) {
        // Filter to specific project
        const match = projects.find((p) => p.id === opts.project);
        if (match) {
          allProjects.push({ parser, project: match });
        }
      } else {
        for (const project of projects) {
          allProjects.push({ parser, project });
        }
      }

      projectSpinner.succeed(
        `  ${parser.displayName}: ${projects.length} project${projects.length !== 1 ? "s" : ""} found`,
      );
    } catch {
      projectSpinner.warn(`  ${parser.displayName}: failed to scan`);
    }
  }

  if (allProjects.length === 0) {
    console.log(chalk.yellow("\n  No projects with chat history found."));
    process.exit(0);
  }

  // Step 3: --list mode — just show what's available
  if (opts.list) {
    console.log();
    console.log(chalk.bold("  Available IDE Chat History"));
    console.log();

    let lastIde = "";
    for (const { parser, project } of allProjects) {
      if (parser.id !== lastIde) {
        console.log(chalk.blue(`  ${parser.displayName}`));
        lastIde = parser.id;
      }
      const date = chalk.dim(project.lastModified.toISOString().slice(0, 10));
      const ws = project.workspacePath ? chalk.dim(` → ${project.workspacePath}`) : "";
      console.log(`    ${chalk.green(project.id.slice(0, 12))} ${project.name} ${date}${ws}`);
    }
    console.log();
    console.log(chalk.dim(`  ${allProjects.length} total project(s) across ${parsers.length} IDE(s)`));
    return;
  }

  // Step 4: Parse conversations from each project
  console.log();
  let totalConversations = 0;
  let totalMessages = 0;
  const importQueue: Array<{
    parser: IDEParser;
    project: IDEProject;
    conversation: ParsedConversation;
  }> = [];

  for (const { parser, project } of allProjects) {
    const parseSpinner = ora(`  Parsing ${parser.displayName} / ${project.name}...`).start();
    try {
      let conversations = await parser.parseConversations(project.id);

      // Filter by date if --since specified
      if (sinceDate) {
        conversations = conversations.filter((c) => {
          if (c.startedAt) return c.startedAt >= sinceDate;
          if (c.endedAt) return c.endedAt >= sinceDate;
          return true; // Include if no date info
        });
      }

      // Apply limit
      conversations = conversations.slice(0, limit);

      for (const conv of conversations) {
        importQueue.push({ parser, project, conversation: conv });
        totalMessages += conv.messages.length;
      }
      totalConversations += conversations.length;

      parseSpinner.succeed(
        `  ${parser.displayName} / ${project.name}: ${conversations.length} conversation${conversations.length !== 1 ? "s" : ""} (${conversations.reduce((sum, c) => sum + c.messages.length, 0)} messages)`,
      );
    } catch {
      parseSpinner.warn(`  ${parser.displayName} / ${project.name}: failed to parse`);
    }
  }

  if (importQueue.length === 0) {
    console.log(chalk.yellow("\n  No conversations to import."));
    return;
  }

  // Step 5: Show summary
  console.log();
  console.log(chalk.bold("  Import Summary"));
  console.log(`  ${chalk.green(String(totalConversations))} conversations, ${chalk.green(String(totalMessages))} messages`);
  console.log(`  Each conversation will be AI-summarized and stored with embeddings`);
  if (opts.rembProject) {
    console.log(`  Target project: ${chalk.blue(opts.rembProject)}`);
  }

  // Step 6: Dry run — stop here
  if (opts.dryRun) {
    console.log();
    console.log(chalk.yellow("  Dry run — no data sent. Remove --dry-run to import."));

    // Show a preview of the first few conversations
    const preview = importQueue.slice(0, 5);
    console.log();
    for (const { parser, conversation } of preview) {
      const title = conversation.title ?? conversation.messages[0]?.text.slice(0, 80);
      const date = conversation.startedAt?.toISOString().slice(0, 10) ?? "unknown";
      console.log(`  ${chalk.dim(date)} ${chalk.cyan(parser.displayName)} ${title}`);
    }
    if (importQueue.length > 5) {
      console.log(chalk.dim(`  ... and ${importQueue.length - 5} more`));
    }
    return;
  }

  // Step 7: Import — send to API in batches
  if (!opts.all) {
    console.log();
    console.log(chalk.dim("  Use --all to skip confirmation, or --dry-run to preview."));
    // In a non-interactive context we proceed; in a future version we could add readline confirmation
  }

  const client = createApiClient();
  let imported = 0;
  let skipped = 0;
  let failed = 0;

  const importSpinner = ora(`  Importing 0/${importQueue.length}...`).start();

  // Process in batches
  for (let i = 0; i < importQueue.length; i += BATCH_SIZE) {
    const batch = importQueue.slice(i, i + BATCH_SIZE);

    const batchResults = await Promise.allSettled(
      batch.map(async ({ parser, project, conversation }) => {
        // Convert parsed conversation → RawConversationEvent[]
        const events = conversationToEvents(conversation);
        if (events.length === 0) return "skipped";

        // Cap events per request
        const cappedEvents = events.slice(0, MAX_EVENTS_PER_REQUEST);

        const result = await client.logSmartConversation({
          events: cappedEvents,
          projectSlug: opts.rembProject ?? undefined,
          ideSource: parser.id,
          metadata: {
            import_source: parser.id,
            import_project_name: project.name,
            import_workspace_path: project.workspacePath,
            conversation_id: conversation.id,
            conversation_title: conversation.title,
            started_at: conversation.startedAt?.toISOString(),
            message_count: conversation.messages.length,
          },
        });

        return result.deduplicated ? "deduplicated" : "imported";
      }),
    );

    for (const result of batchResults) {
      if (result.status === "fulfilled") {
        if (result.value === "skipped" || result.value === "deduplicated") {
          skipped++;
        } else {
          imported++;
        }
      } else {
        failed++;
      }
    }

    importSpinner.text = `  Importing ${Math.min(i + BATCH_SIZE, importQueue.length)}/${importQueue.length}... (${imported} new, ${skipped} skipped)`;
  }

  importSpinner.succeed(
    `  Import complete: ${chalk.green(String(imported))} imported, ${chalk.yellow(String(skipped))} skipped/deduplicated, ${failed > 0 ? chalk.red(String(failed)) : "0"} failed`,
  );

  if (imported > 0) {
    console.log();
    console.log(chalk.dim("  View imported history: remb history"));
    if (!opts.rembProject) {
      console.log(chalk.dim("  Tip: Use --remb-project <slug> to associate imports with a project"));
    }
  }
}

File: packages/cli-go/cmd/import.go
package cmd

import (
	"fmt"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/spf13/cobra"
	"github.com/useremb/remb/internal/api"
	"github.com/useremb/remb/internal/config"
	"github.com/useremb/remb/internal/parsers"
)

var (
	importIDE         string
	importProject     string
	importRembProject string
	importAll         bool
	importDryRun      bool
	importSince       string
	importList        bool
	importLimit       int
)

var importCmd = &cobra.Command{
	Use:   "import",
	Short: "Import AI chat history from local IDE storage into Remb",
	Long: `Import AI chat history from local IDE storage into Remb.

Supported IDEs:
  cursor          Cursor (VS Code fork)
  claude-code     Claude Code CLI
  vscode          VS Code (GitHub Copilot)
  windsurf        Windsurf (Codeium)
  intellij        IntelliJ IDEA
  pycharm         PyCharm
  android-studio  Android Studio
  visual-studio   Visual Studio (Windows)
  zed             Zed
  sublime-text    Sublime Text (LSP-Copilot)

Examples:
  remb import                          # Auto-detect and import
  remb import --list                   # List detected IDEs and projects
  remb import --ide cursor             # Import from Cursor only
  remb import --all --dry-run          # Preview what would be imported
  remb import --since 2025-01-01       # Only import recent conversations`,
	RunE: runImport,
}

func init() {
	importCmd.Flags().StringVar(&importIDE, "ide", "", "Import from a specific IDE only")
	importCmd.Flags().StringVar(&importProject, "project", "", "Import a specific project/workspace by ID")
	importCmd.Flags().StringVar(&importRembProject, "remb-project", "", "Associate imports with this Remb project")
	importCmd.Flags().BoolVar(&importAll, "all", false, "Import all without prompting")
	importCmd.Flags().BoolVar(&importDryRun, "dry-run", false, "Show what would be imported")
	importCmd.Flags().StringVar(&importSince, "since", "", "Only import after this date (YYYY-MM-DD)")
	importCmd.Flags().BoolVar(&importList, "list", false, "List detected IDEs and projects")
	importCmd.Flags().IntVarP(&importLimit, "limit", "l", 100, "Max conversations per project")
}

const (
	batchSize       = 20
	maxEventsPerReq = 100
	dim             = "\033[2m"
	bold            = "\033[1m"
	green           = "\033[32m"
	yellow          = "\033[33m"
	blue            = "\033[34m"
	red             = "\033[31m"
	reset           = "\033[0m"
	checkMark       = "\033[32m✔\033[0m"
)

type projectEntry struct {
	parser  parsers.IDEParser
	project parsers.IDEProject
}

type importItem struct {
	parser  parsers.IDEParser
	project parsers.IDEProject
	conv    parsers.ParsedConversation
}

func runImport(cmd *cobra.Command, args []string) error {
	// Validate options
	if importSince != "" {
		if _, err := time.Parse("2006-01-02", importSince); err != nil {
			return fmt.Errorf("invalid --since format %q, expected YYYY-MM-DD", importSince)
		}
	}
	if importLimit > 500 {
		importLimit = 500
	}

	// Step 1: Detect IDEs
	fmt.Fprintf(os.Stderr, "  Detecting installed IDEs...")

	var ideParsers []parsers.IDEParser
	if importIDE != "" {
		p := parsers.GetParser(parsers.IDESource(importIDE))
		if p == nil {
			allIDs := make([]string, 0)
			for _, pp := range parsers.AllParsers() {
				allIDs = append(allIDs, string(pp.ID()))
			}
			return fmt.Errorf("unknown IDE %q. Supported: %s", importIDE, strings.Join(allIDs, ", "))
		}
		ok, _ := p.Detect()
		if !ok {
			return fmt.Errorf("%s storage not found on this machine", p.DisplayName())
		}
		ideParsers = []parsers.IDEParser{p}
	} else {
		ideParsers = parsers.DetectAvailableIDEs()
	}

	if len(ideParsers) == 0 {
		fmt.Fprintf(os.Stderr, "\n  %sNo supported IDEs detected%s\n", red, reset)
		return nil
	}

	names := make([]string, len(ideParsers))
	for i, p := range ideParsers {
		names[i] = p.DisplayName()
	}
	fmt.Fprintf(os.Stderr, "\r  %s Found %d IDE(s): %s\n", checkMark, len(ideParsers), strings.Join(names, ", "))

	// Step 2: List projects
	var allProjects []projectEntry
	for _, p := range ideParsers {
		projects, err := p.ListProjects()
		if err != nil {
			fmt.Fprintf(os.Stderr, "  %s %s: failed to scan\n", yellow+"⚠"+reset, p.DisplayName())
			continue
		}
		for _, proj := range projects {
			if importProject != "" && proj.ID != importProject {
				continue
			}
			allProjects = append(allProjects, projectEntry{parser: p, project: proj})
		}
		fmt.Fprintf(os.Stderr, "  %s %s: %d project(s)\n", checkMark, p.DisplayName(), len(projects))
	}

	if len(allProjects) == 0 {
		fmt.Fprintf(os.Stderr, "\n  %sNo projects with chat history found.%s\n", yellow, reset)
		return nil
	}

	// Step 3: --list mode
	if importList {
		fmt.Fprintln(os.Stderr)
		fmt.Fprintf(os.Stderr, "  %sAvailable IDE Chat History%s\n\n", bold, reset)

		lastIDE := ""
		for _, pe := range allProjects {
			id := string(pe.parser.ID())
			if id != lastIDE {
				fmt.Fprintf(os.Stderr, "  %s%s%s\n", blue, pe.parser.DisplayName(), reset)
				lastIDE = id
			}
			date := pe.project.LastModified.Format("2006-01-02")
			ws := ""
			if pe.project.WorkspacePath != "" {
				ws = dim + " → " + pe.project.WorkspacePath + reset
			}
			short := pe.project.ID
			if len(short) > 12 {
				short = short[:12]
			}
			fmt.Fprintf(os.Stderr, "    %s%s%s %s %s%s%s%s\n", green, short, reset, pe.project.Name, dim, date, reset, ws)
		}
		fmt.Fprintf(os.Stderr, "\n  %s%d total project(s) across %d IDE(s)%s\n", dim, len(allProjects), len(ideParsers), reset)
		return nil
	}

	// Step 4: Parse conversations
	fmt.Fprintln(os.Stderr)
	var queue []importItem
	totalMessages := 0

	sinceDate := time.Time{}
	if importSince != "" {
		sinceDate, _ = time.Parse("2006-01-02", importSince)
	}

	for _, pe := range allProjects {
		convs, err := pe.parser.ParseConversations(pe.project.ID)
		if err != nil {
			fmt.Fprintf(os.Stderr, "  %s %s / %s: failed to parse\n", yellow+"⚠"+reset, pe.parser.DisplayName(), pe.project.Name)
			continue
		}

		// Filter by --since
		if !sinceDate.IsZero() {
			filtered := make([]parsers.ParsedConversation, 0, len(convs))
			for _, c := range convs {
				if c.StartedAt != nil && c.StartedAt.Before(sinceDate) {
					continue
				}
				filtered = append(filtered, c)
			}
			convs = filtered
		}

		// Apply limit
		if len(convs) > importLimit {
			convs = convs[:importLimit]
		}

		msgCount := 0
		for _, c := range convs {
			queue = append(queue, importItem{parser: pe.parser, project: pe.project, conv: c})
			msgCount += len(c.Messages)
			totalMessages += len(c.Messages)
		}

		fmt.Fprintf(os.Stderr, "  %s %s / %s: %d conversations (%d messages)\n",
			checkMark, pe.parser.DisplayName(), pe.project.Name, len(convs), msgCount)
	}

	if len(queue) == 0 {
		fmt.Fprintf(os.Stderr, "\n  %sNo conversations to import.%s\n", yellow, reset)
		return nil
	}

	// Step 5: Summary
	fmt.Fprintln(os.Stderr)
	fmt.Fprintf(os.Stderr, "  %sImport Summary%s\n", bold, reset)
	fmt.Fprintf(os.Stderr, "  %s%d%s conversations, %s%d%s messages\n",
		green, len(queue), reset, green, totalMessages, reset)
	if importRembProject != "" {
		fmt.Fprintf(os.Stderr, "  Target project: %s%s%s\n", blue, importRembProject, reset)
	}

	// Step 6: Dry run
	if importDryRun {
		fmt.Fprintln(os.Stderr)
		fmt.Fprintf(os.Stderr, "  %sDry run — no data sent. Remove --dry-run to import.%s\n", yellow, reset)
		limit := 5
		if len(queue) < limit {
			limit = len(queue)
		}
		fmt.Fprintln(os.Stderr)
		for _, item := range queue[:limit] {
			title := item.conv.Title
			if title == "" && len(item.conv.Messages) > 0 {
				title = item.conv.Messages[0].Text
				if len(title) > 80 {
					title = title[:80]
				}
			}
			date := "unknown"
			if item.conv.StartedAt != nil {
				date = item.conv.StartedAt.Format("2006-01-02")
			}
			fmt.Fprintf(os.Stderr, "  %s%s%s %s%s%s %s\n", dim, date, reset, blue, item.parser.DisplayName(), reset, title)
		}
		if len(queue) > 5 {
			fmt.Fprintf(os.Stderr, "  %s... and %d more%s\n", dim, len(queue)-5, reset)
		}
		return nil
	}

	// Step 7: Import
	// Resolve project slug
	projectSlug := importRembProject
	if projectSlug == "" {
		cfg := config.FindProjectConfig("")
		if cfg != nil && cfg.Config.Project != "" {
			projectSlug = cfg.Config.Project
		}
	}

	client, err := api.NewClient()
	if err != nil {
		return err
	}

	var (
		imported int
		skipped  int
		failed   int
		mu       sync.Mutex
	)

	fmt.Fprintf(os.Stderr, "\n  Importing 0/%d...", len(queue))

	for i := 0; i < len(queue); i += batchSize {
		end := i + batchSize
		if end > len(queue) {
			end = len(queue)
		}
		batch := queue[i:end]

		var wg sync.WaitGroup
		for _, item := range batch {
			wg.Add(1)
			go func(it importItem) {
				defer wg.Done()

				events := parsers.ConversationToEvents(it.conv)
				if len(events) == 0 {
					mu.Lock()
					skipped++
					mu.Unlock()
					return
				}

				if len(events) > maxEventsPerReq {
					events = events[:maxEventsPerReq]
				}

				// Convert to API types
				smartEvents := make([]api.SmartEvent, len(events))
				for j, e := range events {
					smartEvents[j] = api.SmartEvent{
						Type:      e.Type,
						Text:      e.Text,
						Path:      e.Path,
						Name:      e.Name,
						Timestamp: e.Timestamp,
					}
				}

				meta := map[string]interface{}{
					"import_source":         string(it.parser.ID()),
					"import_project_name":   it.project.Name,
					"import_workspace_path": it.project.WorkspacePath,
					"conversation_id":       it.conv.ID,
					"conversation_title":    it.conv.Title,
					"message_count":         len(it.conv.Messages),
				}
				if it.conv.StartedAt != nil {
					meta["started_at"] = it.conv.StartedAt.Format(time.RFC3339)
				}

				resp, err := client.LogSmartConversation(api.LogSmartConversationRequest{
					Events:      smartEvents,
					ProjectSlug: projectSlug,
					IDESource:   string(it.parser.ID()),
					Metadata:    meta,
				})

				mu.Lock()
				if err != nil {
					failed++
				} else if resp.Deduplicated {
					skipped++
				} else {
					imported++
				}
				mu.Unlock()
			}(item)
		}
		wg.Wait()

		fmt.Fprintf(os.Stderr, "\r  Importing %d/%d... (%d new, %d skipped)", end, len(queue), imported, skipped)
	}

	failStr := "0"
	if failed > 0 {
		failStr = fmt.Sprintf("%s%d%s", red, failed, reset)
	}
	fmt.Fprintf(os.Stderr, "\r  %s Import complete: %s%d%s imported, %s%d%s skipped, %s failed    \n",
		checkMark, green, imported, reset, yellow, skipped, reset, failStr)

	if imported > 0 {
		fmt.Fprintln(os.Stderr)
		fmt.Fprintf(os.Stderr, "  %sView imported history: remb history%s\n", dim, reset)
		if importRembProject == "" {
			fmt.Fprintf(os.Stderr, "  %sTip: Use --remb-project <slug> to associate imports%s\n", dim, reset)
		}
	}

	return nil
}
